### ISSUE TITLE
Cryptic failure: "povray exited with code 139" (EDIT: actually, it's apparently a segmentation fault).

### ISSUE THREAD

**Issue by nlburgin** [POV-Ray/povray#314]
I don't know what this means. The output text had no indication of what it meant.

```
/neil/.jedit/console/commando/> /usr/local/bin/povray +W1920  +H1080 +A  +Q11  +FN -RVP -D /neil/povray/PLUGL.pov
povray: cannot open the user configuration file /home/neil/.povray/3.7/povray.conf: No such file or directory
Persistence of Vision(tm) Ray Tracer Version 3.7.1-rc.1.unofficial (g++ 5.4.0 @
 x86_64-pc-linux-gnu)
This is an unofficial version compiled by:
 Neil Burgin
 The POV-Ray Team is not responsible for supporting this version.

POV-Ray is based on DKBTrace 2.12 by David K. Buck & Aaron A. Collins
Copyright 1991-2017 Persistence of Vision Raytracer Pty. Ltd.

Primary POV-Ray 3.7 Architects/Developers: (Alphabetically)
  Chris Cason         Thorsten Froehlich  Christoph Lipka   

With Assistance From: (Alphabetically)
  Nicolas Calimet     Jerome Grimbert     James Holsenback    Christoph Hormann 
  Nathan Kopp         Juha Nieminen       William F. Pokorny

Past Contributors: (Alphabetically)
  Steve Anger         Eric Barish         Dieter Bayer        David K. Buck     
  Nicolas Calimet     Chris Cason         Aaron A. Collins    Chris Dailey      
  Steve Demlow        Andreas Dilger      Alexander Enzmann   Dan Farmer        
  Thorsten Froehlich  Mark Gordon         Jerome Grimbert     James Holsenback  
  Christoph Hormann   Mike Hough          Chris Huff          Kari Kivisalo     
  Nathan Kopp         Lutz Kretzschmar    Christoph Lipka     Jochen Lippert    
  Pascal Massimino    Jim McElhiney       Douglas Muir        Juha Nieminen     
  Ron Parker          William F. Pokorny  Bill Pulver         Eduard Schwan     
  Wlodzimierz Skiba   Robert Skinner      Yvo Smellenbergh    Zsolt Szalavari   
  Scott Taylor        Massimo Valentini   Timothy Wegner      Drew Wells        
  Chris Young       

Other contributors are listed in the documentation.

Support libraries used by POV-Ray:
  ZLib 1.2.11, Copyright 1995-2012 Jean-loup Gailly and Mark Adler
  LibPNG 1.6.31, Copyright 1998-2012 Glenn Randers-Pehrson
  LibJPEG 6b, Copyright 1991-2016 Thomas G. Lane, Guido Vollbeding
  LibTIFF 4.0.8, Copyright 1988-1997 Sam Leffler, 1991-1997 SGI
  Boost 1.62, http://www.boost.org/

Dynamic optimizations:
  CPU detected: AMD,SSE2,AVX,FMA3,FMA4
  Noise generator: avxfma4-amd.2 (hand-optimized by AMD, 2017-04 update)

Parser Options
  Input file: /neil/povray/PLUGL.pov
  Remove bounds........On 
  Split unions.........Off
  Library paths:
    /usr/local/share/povray-3.7
    /usr/local/share/povray-3.7/ini
    /usr/local/share/povray-3.7/include
  Clock value:    0.000  (Animation off)
Image Output Options
  Image resolution.....1920 by 1080 (rows 1 to 1080, columns 1 to 1920).
  Output file........../neil/povray/PLUGL.png, 24 bpp PNG
  Dithering............Off
  Graphic display......Off
  Mosaic preview.......Off
  Continued trace......Off
Information Output Options
  All Streams to console..........On 
  Debug Stream to console.........On 
  Fatal Stream to console.........On 
  Render Stream to console........On 
  Statistics Stream to console....On 
  Warning Stream to console.......On 
==== [Parsing...] ==========================================================
Possible Parse Error: assumed_gamma not specified in this POV-Ray 3.7 or later
 scene. Future versions of POV-Ray may consider this a fatal error. To avoid
 this warning, explicitly specify 'assumed_gamma 1.0' in the global_settings
 section. See the documentation for more details.
----------------------------------------------------------------------------
Parser Statistics
----------------------------------------------------------------------------
Finite Objects:            2
Infinite Objects:          0
Light Sources:             1
Total:                     3
----------------------------------------------------------------------------
Parser Time
  Parse Time:       0 hours  0 minutes  0 seconds (0.025 seconds)
              using 1 thread(s) with 0.020 CPU-seconds total
  Bounding Time:    0 hours  0 minutes  0 seconds (0.000 seconds)
              using 1 thread(s) with 0.000 CPU-seconds total
----------------------------------------------------------------------------
Render Options
  Quality:  9
  Bounding boxes.......On   Bounding threshold: 3
  Antialiasing.........On  (Method 1, Threshold 0.300, Depth 3, Jitter 1.00,
 Gamma 2.50)
==== [Rendering...] ========================================================
Process /usr/local/bin/povray exited with code 139
```

According to the manual, there shouldn't even be an exit code 139. Just code 0 for normal exit, code 1 for any fatal error, and code 2 for user abort. Neither the manual nor a Google search turned up any reference to Pov-Ray exit code 139.

All I know is it didn't do this until I added photon mapping to my scene.

```
#version 3.7;
global_settings {
    radiosity {
      pretrace_start 0.08
      pretrace_end   0.01
      count 150
      nearest_count 10
      error_bound 0.5
      recursion_limit 3
      low_error_factor 0.5
      gray_threshold 0.0
      minimum_reuse 0.005
      maximum_reuse 0.2
      brightness 1
      adc_bailout 0.005
    }
    max_trace_level 50
    photons {
    	count 40000
    	autostop 0
    }
}
#include "colors.inc"    // The include files contain
#include "stones.inc"    // pre-defined scene elements
#include "textures.inc"    // pre-defined scene elements
#include "shapes.inc"
#include "glass.inc"
#include "metals.inc"         
#include "woods.inc"
#include "functions.inc"


camera {
	up<0,1,0>
	right<1.77,0,0>
    location <0, 2, -6>
    look_at  <0, 0.5,  0.5>    
}

isosurface{
	 function {
	   f_sphere(x,y,z,2) + (f_noise3d(x*10,y*10,z*10)*4)
	 }
	 contained_by {box {1,-1}}
	 max_gradient 45
	 texture { pigment { color Clear } finish { F_Glass1 } }
	 /* texture {
		 pigment {color Yellow }
		 rotate <45,45,0>
		 scale 0.05
     } */
     interior { ior 1.5 }
     photons {
		target
		refraction on
		reflection on
		collect off
     }
}

/* isosurface{
	 function {
	   f_sphere(x,y,z,2) + f_noise3d(x,y,z)
	 }
	 contained_by {box {1,-1}}
	 max_gradient 1.8
	 texture { pigment { color Clear } finish { F_Glass1 } }
} */

/* sphere{
	<0,1,0>, 1
	texture { pigment { color Clear } finish { F_Glass1 } }
	interior { ior 1.5 }
}  */

/* superellipsoid { <.1,1>
    texture { pigment { color Clear } finish { F_Glass1 } }
	interior { ior 1.5 }
  } */


isosurface{
	 function {
	   atan(x*10)+(y*2)
	 }
	 contained_by {box {<-1,-1,-1>,<5,5,2>}}
	 max_gradient 10
	 texture {
		 pigment { square color Red, color White, color Blue, color White }
		 rotate <45,45,0>
		 scale 0.05
     }
     scale 8
     rotate <0,90,0>
     translate <-2.5,2,6>
}

sky_sphere {
  pigment {
    gradient y
      color_map {
        [ 0.5  color White ]
        [ 1.0  color Black ]
        }
    scale 2
    translate -1
    }
  emission rgb <0.8,0.8,1>
  }

/* box {
   5,-5
   texture {
     T_Stone25     // Pre-defined from stones.inc
     scale 4       // Scale by the same amount in all
                   // directions
   }
   rotate x*45     // Equivalent to "rotate <0,20,0>"
} */

light_source { <2, 4, -3> color White
	photons {
		refraction on
		reflection on
    }
}
```

(Sorry the file is such a mess, I'm pretty new to this so I was experimenting).

**Comment by nlburgin** [comment]
Update: I just ran it in a regular command shell instead of the one provided by Jedit, and where it had previously said, ``Process /usr/local/bin/povray exited with code 139`` it instead said ``Segmentation fault``.

So apparently this is actually a segfault; I don't know where the jedit console was getting code 139...

I'm not sure if the segfault is because of a bug in RC1 or if it's just my build... if you can't easily reproduce it using the included scene file, then probably it's just something wrong with my build and you should mark this invalid... 

anyway, I'm going to attach the file, since that's probably more standard than having it embedded in a post.

(and if you're wondering what PLUGL means... it doesn't mean anything. I just didn't bother to make up a sensible name for the file I was experimenting with)


[PLUGL.pov.zip](https://github.com/POV-Ray/povray/files/1227537/PLUGL.pov.zip)

**Comment by wfpokorny** [comment]
A quick run of the PLUGL.pov file using no flags and with my compile of the release/3.8.0 branch ran to completion. I have your command line running now (57%) and looks to have gotten further than your posted results. I have a 3.7.1 () compile too I'll try later, but it is basically 3.8.0 at the moment - other than the re-version to 3.8.0 for release - so I expect it will run too.   

What does the command: 'ulimit -a' show in your environment - both where you compiled and where you ran? I'm wondering if perhaps your system's default stack size setting is too small for POV-Ray - as I remember it currently needs to be at least 4MB by default on *nix systems to get each thread going cleanly. My Ubuntu systems - 32 bit and 64 bit - have been set at a ulimit of 8MB for the stack for a long time by default. 

Stack failures are often strange.

**Comment by nlburgin** [comment]
```
neil@new-host-192-168-1-19 ~ $ ulimit -a
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 31898
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 1024
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 31898
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
```

Looks like the stack size is around 8MB.

Is a stack overflow the same as a segfault? Like, if it happens in a release-optimized build with no backtrace, would it be indistinguishable from array out-of-bounds, use after free, etc? I didn't know that.

Anyway, if you got the file working past where I did on the same settings, then either there is no real bug here, or else it was inadvertently fixed since RC1. Either way, that would mean there's nothing for you to worry about.

I don't deny it could have been my build. I don't remember exactly what my compile settings were, but I ''might'' have tried something crazy like trying to enable Link-Time Optimization. 

Still, I would expect that more to cause a compile failure, than to compile correctly and run for a while but then segfault under very specific conditions...

**Comment by nlburgin** [comment]
I just compiled an unoptimized debug build from the same sources and the segfault didn't happen.

Should have tried that first... Probably time to close this. Sorry.

**Comment by nlburgin** [comment]
Later, I'm going to try a build with default optimizations enabled but no personally customized CXXFLAGS. If that works, then this bug is definitely invalid.

**Comment by nlburgin** [comment]
The default-optimized build also works.

In the end, it was indeed an issue with my build. Nothing more to see here...

**Comment by c-lipka** [comment]
You may want to report this issue to the GCC developers. If their linker screws up programs, I'm sure they want to know.

**Comment by c-lipka** [comment]
> So apparently this is actually a segfault; I don't know where the jedit console was getting code 139...

That's presumably the value jedit got from the kernel. If a program is terminated due to a /signal/ (such as Segmentation Violation, SIGSEGV) rather than normal operation, the kernel reports a faux exit code of 128+N instead, where N is the signal code (11 in case of SIGSEGV on Linux machines).

**Comment by wfpokorny** [comment]
I've been off on other things of late and only checking email/github in the mornings. So, I missed your posts yesterday. Happy you are running OK with a default optimized build. 

The link time optimization option I've tried twice now - at different times - for POV-Ray compiles. It buys quite a lot in our windows builds, but it's never helped at all really on my Ubuntu system for what I can get going in a day or so. I suspect a big part of the 'not helping' comes from the POV-Ray developers over time doing a lot of explicit inlines of code - and this, by my gut guess, is where I think the link time optimization typically buys the most with gcc, lto builds. Again this is me guessing as to why no POV-Ray performance gains with gcc lto. Where I've seen reports with the gcc link time optimization it sometimes buys quite a bit and others like me have found it buying not much. The one real gain for POV-Ray is the resulting load module is about 2/3 the size, which might itself be of value for resource constrained environments. 

Anyway, the link time optimization requires special linkers and archive programs and I had to hand hack the resulting autotool's Makefiles to force their use after installing them. In other words, getting lto to work has been pretty messy to do. I've also never run more than the built in benchmark with such builds, so perhaps there are problems in waiting with such builds.     

Off the top of my head, I think, the code 139 is the result of the shell adding a value to certain 'serious' signals and a segfault is one such beast. One has to subtract, AND - or something - to get the real signal value. A signal value which could be other than 0,1 (or 2?) in cases where POV-Ray's *nix signal handler didn't catch it.

**Comment by nlburgin** [comment]
> it's never helped at all really on my Ubuntu system for what I can get going in a day or so.

AFAIK, that could be caused by "fat LTO objects" which contain both normal object code and LTO bytecode, and if the linker isn't set up properly for LTO, then it will just use the normal object code without actually doing the LTO, and the only difference from a non-lto build will be that it took longer to build. To test this, you could try passing gcc the `-fno-fat-lto-objects` flag and see if it still builds at all. If it doesn't, then the previous attempt wasn't even actually using LTO.

>You may want to report this issue to the GCC developers. If their linker screws up programs, I'm sure they want to know.

Before doing that, I'd probably need to try it again with the latest GCC (I think it's GCC7), instead of GCC5. It could be fixed already.

Also, I don't think it's actually considered terribly unexpected for LTO to screw things up if the build system (makefiles or whatever) wasn't set up to accomodate it. Although, then again, I think that tends to be more in terms of the build failing completely than producing a buggy output, so this I guess *could* still be a gcc/binutils bug.

**Comment by wfpokorny** [comment]
Given you might try lto again, I'll lay out my own questions / ignorance with respect to gcc lto behavior.  

* I did not try  -fno-fat-lto-objects and will put it on my own to-try list for any further attempt. I took the reduced size to indicate the part of the optimization which recognized redundancy (supposing mostly due template use) was being recognized and eliminated. Maybe other lto optimizations didn't really kick in? 

*  I had the thought to get the most out of lto, perhaps all the dependent libraries need to be enabled for lto too? I've made no attempt to rebuild POV-Ray's dependent libraries with lto support to test the thought.  

* A static link(a) of POV-Ray itself buys 0.5-2% in my testing. I didn't try lto with a static link and perhaps this significantly changes its behavior?

As you most probably know, measured performance varies substantially depending upon the system's current cache usage. I've gotten to where I run performance tests on otherwise quiet systems and with POV-Ray threads (+wt<n>) <= to the machine core count. Multiple measurements helpful too.      

(a) - For a static link to work, additional link flags have to be manually added at the moment for libtiff. I was playing with some build system changes optionally using pkgconfig and being aware of whether the link was static or dynamic to fix the './configure --enable-static' linking, but the attempt at a build system update is on hold.
